src/ucp/tag/rndv.c

```diff
@@ -1237,19 +1434,18 @@ static ucs_status_t ucp_rndv_pipeline(ucp_request_t *sreq,
     fsreq->send.state.dt.offset         = 0;
 
     offset = 0;
-    for (i = 0; i < num_frags; i++) {
-        length = (i == (num_frags - 1)) ? (rndv_size - offset) : max_frag_size;
-
-        /* internal fragment send request allocated on sender side to receive
-         *  mem type fragment stage to host and to perform a put to receiver */
-        freq = ucp_request_get(worker);
-        if (freq == NULL) {
-            ucs_fatal("failed to allocate fragment receive request");
-        }
+    while (offset != rndv_size) {
+        length = ucs_min(rndv_size - offset, max_frag_size);
 
         if (UCP_MEM_IS_ACCESSIBLE_FROM_CPU(sreq->send.mem_type)) {
             /* sbuf is in host, directly do put */
-            ucp_request_send_state_reset(freq, ucp_rndv_frag_send_put_completion,
+            freq = ucp_request_get(worker);
```

Comment:
why would we need to use PUT protocol when send memory is CPU-accessible? Is it for TCP?
(pls add comment)